### PR TITLE
[New Contribution] Conflict-Aware Fusion training components (LIRE, RLVF)

### PR DESCRIPTION
**Pull request by 14H034160212**
## New Contribution Check List

- [x] I created a subdirectory named with the format `contrib/<my_github_user_name>-<feature_name>`: `contrib/14H034160212-conflict-aware-fusion/`.
- [x] I included a short `README.md` that describes the contribution, its motivation and status, and how to try/evaluate it.
- [x] I added a `LICENSE`. Follows the repo defaults: Apache-2.0 for code, CC-BY-4.0 for docs.
- [x] Code is in a `conflict-aware-fusion` subdirectory. (No `tests` subdirectory: this is an excerpted reference implementation from a published, tested pipeline — see below.)

## Description of Contribution

This is a documentation + reference-code contribution (status: **Validated / published**, not speculative) of two training components that address a reasoning-robustness gap relevant to any consortium-trained base model.

**Problem:** LLMs exhibit "Logic Inertia" — they keep deducing along a learned reasoning chain even when their own premises are inconsistent, instead of halting. On a controlled diagnostic benchmark (four orthogonal stress tests derived from a single canonical rule system), GPT-4o scores only 0.560 and Gemma-3-4B-IT only 0.439 on contradiction-injection, despite scoring 0.79–0.82 on the base task — the failure is specific to premise-inconsistency handling, not general reasoning capability. Untreated open-weight baselines (BERT/Qwen2/TinyLlama) collapse from 1.00 to 0.00.

**What's contributed:** the LIRE (Logical Invariance REgularisation) and RLVF (Reinforcement Learning from Verification Feedback) training stages from our four-stage Conflict-Aware Fusion pipeline (SFT→DPO→LIRE→RLVF), which brings a Qwen3-8B backbone to 1.000 across all four stress-test splits. Full details, results tables, and the Lean-4 verified extension are in the paper.

Paper: https://arxiv.org/abs/2512.06393 — "Conflict-Aware Fusion: Mitigating Logic Inertia in Large Language Models via Structured Cognitive Priors" (Bao, Fu, Witbrock)
Full code/benchmark: https://github.com/14H034160212/lemo (MIT license)

**Relevance to Tapestry:** a candidate post-training stage to make sovereign derivatives of the shared base model more reliably reject inconsistent inputs — a property useful across any partner's domain.

## Related Issues

None yet — happy to link this against a relevant training/evaluation issue if a maintainer points me to one.

### Code Description

- `stage2_train_dpo.py` — DPO stage sharpening the halt-on-contradiction decision boundary (included for pipeline context).
- `stage3_train_lire.py` — symmetric-KL invariance loss across logically-equivalent rule reformulations.
- `stage4_train_rlvf.py` — RL reward formulation using a deterministic symbolic forward-chaining oracle (not a new optimiser), jointly optimising invariance and contradiction-sensitivity via REINFORCE.
- LoRA fine-tuning against Qwen2/Qwen3-family checkpoints (`transformers.Trainer` + PEFT).

### Testing Performed

These scripts are excerpted as-is from the published, evaluated pipeline (results above and in the paper's Table 2/3). No additional tests were added for this excerpt; full evaluation harness and results are in the linked repository. Local absolute cache paths from the source repo were cleaned up to be portable before including here; all three files were verified to `py_compile` cleanly.

### Example Usage

See each script's module docstring for the expected data schema and a runnable example command (e.g. `python stage3_train_lire.py --model_name <qwen-checkpoint> --lire_lambda 1.0 ...`). The full data-generation pipeline that produces the expected input format is in the linked `lemo` repository.

### CODE REVIEW

**Comment by 14H034160212** [comment]
I recently submitted a dataset contribution proposal and two code contributions to Project Tapestry (this PR and #164). My research focuses on LLM post-training, alignment, logical reasoning, reward modelling, and evaluation.

A bit about me: I'm Qiming Bao, an AI Research Engineer at Xtracta (Auckland, NZ), with a PhD in Computer Science from the University of Auckland, advised by Prof. Michael Witbrock. I've ranked #1 on the ReClor logical-reasoning benchmark and published at ACL, AAAI, and IJCAI. More at https://qiming-bao.pages.dev/

I would also be very interested in contributing as a researcher, particularly to the cultural alignment, evaluation, or post-training workstreams. Please let me know if there are any working groups, research meetings, or additional contribution opportunities that I could join.

**Review by deanwampler — CHANGES_REQUESTED** [review]
Hi. Sorry for the delay in reviewing this PR and thank you for the submission. Could you make a few changes?

- [ ] Merge out the latest updates from `develop`.
- [ ] Add your contribution to the table in [`contrib/README.md`](https://github.com/The-AI-Alliance/tapestry/blob/develop/contrib/README.md).
- [ ] From the project root directory, run the following command:
```shell
make SRC_DIR=contrib/14H034160212-conflict-aware-fusion --include-dir=contrib/14H034160212-conflict-aware-fusion format ruff pylint type-check tests
```

* Do all the targets successfully "build"? This command is run for each contribution as part of `make before-pr`, which has to pass before merging. See the discussion in [`contrib/README.md`](https://github.com/The-AI-Alliance/tapestry/blob/develop/contrib/README.md#the-project-wide-make-processes). 
* For contributions, it is okay to disable any of these "quality" steps you don't care about. For example, to disable the `pylint` and `type-check` targets, create a file in your directory called `.custom.mk` with this content:

```makefile
pylint-default type-check-default:
  @echo "${skip-contrib-target}"
  @true
```

Once these are done, I'll do a final review and approve the PR. Thanks.

**Comment by 14H034160212** [comment]
Thanks for the review, @deanwampler — all set:

- [x] Merged out the latest `develop`.
- [x] Added `14H034160212-conflict-aware-fusion` to the table in [`contrib/README.md`](https://github.com/The-AI-Alliance/tapestry/blob/develop/contrib/README.md).
- [x] `make SRC_DIR=contrib/14H034160212-conflict-aware-fusion --include-dir=contrib/14H034160212-conflict-aware-fusion format ruff pylint type-check tests` now passes. `black` reformatted the three files (fixes committed), `ruff --fix` cleaned up the rest (plus two dead-variable removals it flagged), and I added `.custom.mk` to skip `pylint`/`type-check` for this excerpt, following the pattern you showed — its dependencies (`transformers`, `peft`, `trl`, `pandas`, ...) aren't part of the project's own dependency set, so those two checks aren't meaningful here.

Ready for another look whenever convenient. Thanks again for taking the time to review!

**Review by deanwampler — APPROVED** [review]
Looks good. Thanks!
